infra/discovery/src/listener/tools/consistency/offer.js

```diff
@@ -0,0 +1,99 @@
+const db = {
+  ...require('../../../models')
+}
+
+const log = require('../../logger')
+const { assert, getListenerBlock, getPastEvents } = require('./utils')
+
+async function offerCreatedEvents(contract, fromBlock, toBlock) {
+  if (!toBlock || toBlock === 'latest') {
+    toBlock = await getListenerBlock('main', 'Marketplace_')
+  }
+  return await getPastEvents(contract, 'OfferCreated', {
+    fromBlock,
+    toBlock
+  })
+}
+
+/**
+ * Verify that the DB records are sane on a cursory level
+ *
+ * @param netId {number} The Ethereum network ID we're checking
+ * @param event {object} The event object from getPastEvents()
+ * @returns {bool} whether or not it was validated
+ * @throws {Error} describing the assertion error (offer invalid)
+ */
+async function verifyOfferDBRecord(netId, event) {
+  const listingId = event.returnValues.listingID
+  const offerId = event.returnValues.offerID
+  const fqListingID = `${netId}-000-${listingId}`
+  const fqOfferID = `${fqListingID}-${offerId}`
+
+  const records = await db.Offer.findAll({
+    where: {
+      id: fqOfferID
+    }
+  })
+
+  // Make sure we have the expected amount of records
+  assert(
+    records.length > 0,
+    `Did not find a matching record for offer #${fqOfferID}`
+  )
+  assert(records.length < 2, `Found too many records for offer #${fqOfferID}`)
+
+  const offer = records[0]
+
+  assert(
+    offer.listingId === fqListingID,
+    `Invalid listing ID.  ${offer.listingId} !== ${listingId}`
+  )
+
+  assert(
+    offer.buyerAddress === event.returnValues.party.toLowerCase(),
+    `Unexpected buyer.  ${offer.buyerAddress} !== ${event.returnValues.party}`
+  )
+
+  return true
+}
+
+/**
+ * Verify that offer DB records have been created for every OfferCreated event
+ *
+ * @param args {object} map of function arguments
+ */
+async function validateOffers({ contractsContext, fromBlock }) {
+  const events = await offerCreatedEvents(
```

Comment:
Would be good to test other events. Such as OfferAccepted, OfferFinalized, etc...
Primarily what I'd want to make sure is that in the db offer.status reflect the latest event for that offerId.
e.g. if latest event was OfferAccept, then offer.status should be "Accepted".
ok to add as a TODO that can be implemented once we run current script in prod.

Reply:
Adding TODO for now.  Would be happy to continue to improve this.